Testnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs

```diff
@@ -467,6 +489,103 @@ private void CanTransfer(ulong tokenId)
     /// <param name="tokenId">ID of the NFT to validate.</param>
     private void ValidNFToken(ulong tokenId)
     {
-        Assert(GetIdToOwner(tokenId) != Address.Zero);
+        Address tokenOwner = GetIdToOwner(tokenId);
+        EnsureAddressIsNotEmpty(tokenOwner);
+    }
+
+    /// <summary>
+    /// Sets the contract owner who can mint/burn
+    /// </summary>
+    /// <param name="owner"></param>
+    public void TransferOwnership(Address owner)
+    {
+        EnsureOwnerOnly();
+        EnsureAddressIsNotEmpty(owner);
+
+        Log(new OwnershipTransferedLog { PreviousOwner = Owner, NewOwner = owner });
+
+        Owner = owner;
+    }
+
+    private void EnsureOwnerOnly()
+    {
+        Assert(Message.Sender == Owner, "Only owner of the contract can set new owner.");
+    }
+
+    /// <summary>
+    /// Mints new tokens
+    /// </summary>
+    /// <param name="to">The address that will own the minted NFT</param>
+    public void Mint(Address to)
+    {
+        var tokenId = checked(++TokenIdCounter);
+
+        Mint(to, tokenId);
+    }
+
+    /// <summary>
+    /// Mints new tokens
+    /// </summary>
+    /// <param name="to">The address that will own the minted NFT</param>
+    /// <param name="data">the data param will passed destination contract</param>
+    public void SafeMint(Address to, byte[] data)
+    {
+        var tokenId = checked(++TokenIdCounter);
+
+        Mint(to, tokenId);
+
+        EnsureContractReceivedToken(Address.Zero, to, tokenId, data);
+    }
+
+    private void Mint(Address to, ulong tokenId)
+    {
+        if (OwnerOnlyMinting)
+        {
+            EnsureOwnerOnly();
+        }
+
+        EnsureAddressIsNotEmpty(to);
+
+        AddNFToken(to, tokenId);
+
+        LogTransfer(Address.Zero, to, tokenId);
+    }
+
+    /// <summary>
+    /// Notify contract for received token if destination(to) address is a contract.
+    /// Raise exception if notification fails. 
+    /// </summary>
+    /// <param name="from">The address which previously owned the token.</param>
+    /// <param name="to">Destination address that will receive the token</param>
+    /// <param name="tokenId">The token identifier which is being transferred.</param>
+    /// <param name="data">Additional data with no specified format.</param>
+    private void EnsureContractReceivedToken(Address from, Address to, ulong tokenId, byte[] data)
+    {
+        if (State.IsContract(to))
+        {
+            var result = Call(to, 0, "OnNonFungibleTokenReceived", new object[] { Message.Sender, from, tokenId, data }, 0);
+            Assert((bool)result.ReturnValue);
+        }
+    }
```

Comment:
I would add a re-entrancy lock on any methods that involve this method.

Lock the contract in the beginning of any public method making an external call to any contract not directly trusted or created by this one. 
Unlock the contract after the method is finished.

Example from out contracts: 
https://github.com/Opdex/opdex-v1-core/blob/main/src/Contracts/Pools/OpdexStakingPool.cs#L192-#L207

Reply:
Re-entrancy is not an attack or invalid case that should to be prevented. Re-entrancy is a perfectly valid case which should not be prevented. Methods should safely handle re-entrancy cases. 

I wouldn't recommended your lock mechanism because any exception raise between of EnsureUnlocked and UnLock methods will be locked forever. Also as i said you prevent valid use cases in this way. 

There are 2 place where EnsureContractReceivedToken called and there is no any kind of state modification after this method called so there is no security issue on contract.